pkg/networkdevice/diagnoses/diagnoses.go

```diff
@@ -0,0 +1,73 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023-present Datadog, Inc.
+
+// Package diagnoses implements the diagnosis collection for NDM resources
+package diagnoses
+
+import (
+	"fmt"
+	"github.com/DataDog/datadog-agent/pkg/diagnose/diagnosis"
+	"github.com/DataDog/datadog-agent/pkg/networkdevice/metadata"
+)
+
+// Diagnoses hold diagnoses for a NDM resource
+type Diagnoses struct {
+	resourceType string
+	resourceID   string
+	diagnoses    []metadata.Diagnosis
+}
+
+var severityMap = map[string]diagnosis.Result{
+	"success": diagnosis.DiagnosisSuccess,
+	"error":   diagnosis.DiagnosisFail,
+	"warn":    diagnosis.DiagnosisWarning,
+}
+
+// NewDeviceDiagnoses returns a new Diagnoses for a NDM device resource
+func NewDeviceDiagnoses(deviceID string) *Diagnoses {
+	return &Diagnoses{
+		resourceType: "device",
+		resourceID:   deviceID,
+	}
+}
+
+// Add adds a diagnoses
+func (d *Diagnoses) Add(result string, code string, message string) {
+	d.diagnoses = append(d.diagnoses, metadata.Diagnosis{
+		Severity: result,
+		Code:     code,
+		Message:  message,
+	})
+}
+
+// Report returns diagnosis metadata
+func (d *Diagnoses) Report() []metadata.DiagnosisMetadata {
+
+	if d.diagnoses == nil {
+		return []metadata.DiagnosisMetadata{{ResourceType: d.resourceType, ResourceID: d.resourceID, Diagnoses: []metadata.Diagnosis{}}}
+	}
+
+	return []metadata.DiagnosisMetadata{{ResourceType: d.resourceType, ResourceID: d.resourceID, Diagnoses: d.diagnoses}}
+}
+
+// Reset clears current diagnoses
+func (d *Diagnoses) Reset() {
+	d.diagnoses = nil
+}
+
+// ConvertToCLI converts diagnoses to diagnose CLI format
+func (d *Diagnoses) ConvertToCLI() []diagnosis.Diagnosis {
+	var cliDiagnoses []diagnosis.Diagnosis
+
+	for _, diag := range d.diagnoses {
```

Comment:
Topic 1/ Maybe could be `ReportAsAgentDiagnoses` or `ReportAsCliDiagnoses`.


Topic 2/ I think for CLI output purpose we might need to save the prevDiagnoses (for different reason than the [discussion here](https://github.com/DataDog/datadog-agent/pull/18973#discussion_r1307670161))

With the currently implementation, there is possibly a concurrency issue with Diagnoses Reset/Clear and the call to `ConvertToCLI()`.

For example:

1/ Check run starts
2/ Diagnoses Reset called.
3/ `ConvertToCLI()` Is called with Diagnoses empty or partially filled.
4/ Check run finishes

As you can see in the example above, if `ConvertToCLI()` is called when the integration is running, it can result in wrong diagnoses being reporting via CLI.

Maybe to fix this, a possible solution:
a/ Use Diagnoses.diagnoses to track diagnoses during check run
b/ When `Report()` is called
  - flush/copy `Diagnoses.diagnoses` to `Diagnoses.lastFlushedDiagnoses`
  - clear `Diagnoses.diagnoses`
  - Report `Diagnoses.lastFlushedDiagnoses`
c/ When `ConvertToCLI()` use `Diagnoses.lastFlushedDiagnoses` to build the output.
d/ note: we don't need `Reset()` anymore since `Report()` already does that job